docs/dev-guide.md

```diff
@@ -123,85 +83,81 @@ In general we prefer simplicity. We standardize on JavaScript (Node.js) and SQL
+- `instructorGradebook.ts` exports an Express `Router` instance and has the following basic structure:
 
   ```javascript
-  var ERR = require('async-stacktrace');
-  var _ = require('lodash');
-  var express = require('express');
-  var router = express.Router();
-  var sqldb = require('@prairielearn/postgres');
-  var sql = sqldb.loadSqlEquiv(__filename);
-
-  router.get('/', function (req, res, next) {
-    var params = { course_instance_id: res.params.courseInstanceId };
-    sqldb.query(sql.user_scores, params, function (err, result) {
-      // SQL queries for page data
-      if (ERR(err, next)) return;
-      res.locals.user_scores = result.rows; // store the data in res.locals
-
-      res.render('pages/instructorGradebook/instructorGradebook', res.locals); // render the page
-      // inside the EJS template, "res.locals.var" can be accessed with just "var"
-    });
-  });
+  import { Router } from 'express';
+  import asyncHandler from 'express-async-handler';
 
-  module.exports = router;
-  ```
+  import { loadSqlEquiv, queryRows } from '@prairielearn/postgres';
 
-- Use the `res.locals` variable to build up data for the page rendering. Many basic objects are already included from the `selectAndAuthz*.js` middleware that runs before most page loads.
+  import { InstructorGradebook, UserScoresRowSchema } from './instructorGradebook.html.js';
 
-- Use [EJS templates](http://ejs.co) (Embedded JavaScript) templates for all pages. Using JS as the templating language removes the need for another ad hoc language, but does require some discipline to not get in a mess. Try and minimize the amount of JS code in the template files. Inside a template the JS code can directly access the contents of the `res.locals` object.
+  const router = Router();
+  const sql = loadSqlEquiv(import.meta.url);
 
-- Sub-templates are stored in `pages/partials` and can be loaded as below. The sub-template can also access `res.locals` as its base scope, and can also accept extra arguments with an arguments object:
+  router.get(
+    '/',
+    asyncHandler(async (req, res, next) => {
+      const rows = await queryRows(
+        sql.select_user_scores,
+        {
+          course_instance_id: res.params.courseInstanceId,
+        },
+        UserScoresRowSchema,
+      );
+
+      res.send(InstructorGradebook({ rows, resLocals: res.locals }));
+    }),
+  );
 
-  ```javascript
-  <%- include('../partials/assessment', {assessment: assessment}); %>
+  export default router;
   ```
 
+- When possible, prefer explicitly passing individual typed properties to templates instead of adding properties to `res.locals`. However, `res.locals` may be used for data coming from middlewares that will be used on many pages.
+
+- Use [`@prairielearn/html`](https://github.com/PrairieLearn/PrairieLearn/tree/master/packages/html) to generate HTML pages. This uses HTML tagged-template literals to generate HTML, which in turn makes it easy to get full type-checking.
```

Comment:
Do you want to include an example of html.ts file, similar to the router ts above?

Reply:
That makes sense. It might be worth using a relatively simple example in that case, I don't think the gradebook is a good idea.

Reply:
I agree, especially since it's not very modern.

Reply:
Done in a86dbc5.